sql/handler.cc

```diff
@@ -6791,116 +6815,137 @@ bool Vers_parse_info::check_and_fix_implicit(
+    !vers_info.with_system_versioning)
   {
     // All is correct but this table is not versioned.
-    create_info->options&= ~HA_VERSIONED_TABLE;
+    options&= ~HA_VERSIONED_TABLE;
     return false;
   }
 
-  if ((system_time.start || system_time.end || as_row.start || as_row.end) &&
-      !with_system_versioning)
+  if (!vers_info.with_system_versioning && vers_info.any_sys_field_declared())
   {
-    my_error(ER_MISSING, MYF(0), table_name, "WITH SYSTEM VERSIONING");
+    my_error(ER_MISSING, MYF(0), create_table.table_name, "WITH SYSTEM VERSIONING");
     return true;
   }
 
-  TABLE *orig_table= NULL;
+  if (vers_tables)
+  {
+    DBUG_ASSERT(options & HA_VERSIONED_TABLE);
+    DBUG_ASSERT(versioned_write);
+    *versioned_write= true;
+  }
+
   List_iterator<Create_field> it(alter_info->create_list);
+  int added= 0;
   while (Create_field *f= it++)
   {
-    if (is_trx_start(*f))
+    if ((f->versioning == Column_definition::VERSIONING_NOT_SET &&
+         !vers_info.with_system_versioning) ||
+        f->versioning == Column_definition::WITHOUT_VERSIONING)
+    {
+      f->flags|= VERS_UPDATE_UNVERSIONED_FLAG;
+    }
+
+    if (!vers_tables)
+      continue;
+
+    DBUG_ASSERT(versioned_write);
+    if (vers_info.is_start(*f) &&
+      vers_info.default_start == f->field_name)
     {
-      if (!as_row.start) // not inited in CREATE ... SELECT
+      if (vers_info.as_row.start)
+        it.remove();
+      else
       {
-        DBUG_ASSERT(vers_tables > 0);
-        if (orig_table && orig_table != f->field->orig_table)
-        {
-          err_different_tables:
-          my_error(ER_VERS_DIFFERENT_TABLES, MYF(0), table_name);
-          return true;
-        }
-        orig_table= f->field->orig_table;
-        as_row.start= f->field_name;
-        system_time.start= as_row.start;
+        vers_info.set_start(f->field_name);
+        *versioned_write= false;
       }
       continue;
     }
-    if (is_trx_end(*f))
+    if (vers_info.is_end(*f) &&
+      vers_info.default_end == f->field_name)
     {
-      if (!as_row.end)
+      if (vers_info.as_row.end)
+        it.remove();
+      else
       {
-        DBUG_ASSERT(vers_tables > 0);
-        if (orig_table && orig_table != f->field->orig_table)
-        {
-          goto err_different_tables;
-        }
-        orig_table= f->field->orig_table;
-        as_row.end= f->field_name;
-        system_time.end= as_row.end;
+        vers_info.set_end(f->field_name);
+        *versioned_write= false;
       }
       continue;
     }
 
-    if ((f->versioning == Column_definition::VERSIONING_NOT_SET &&
-         !with_system_versioning) ||
-        f->versioning == Column_definition::WITHOUT_VERSIONING)
+    uint flags_left= VERS_SYSTEM_FIELD;
+    if (flags_left && (vers_info.is_start(*f) || vers_info.is_end(*f)))
     {
-      f->flags|= VERS_UPDATE_UNVERSIONED_FLAG;
-    }
-  }
+      uint sys_flag= f->flags & flags_left;
+      flags_left^= sys_flag;
```

Comment:
It's a dead store. `flags_left` looks suspicious.

Reply:
It's a typo. Should be:
```c++
flags_left-= sys_flag;
```
Thanks!